src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,173 @@
+        */
+        T append(T newValue, bool& isValid);
+
+        /**
+        * :return unordered content (data ) of this sl. window; call getLinearizedData() if you need them oredered from oldest->newest
+        * This direct access method is fast.
+        */
+        const std::vector<T>& getData() const;
+
+        /** linearize method for the internal buffer; this is slower than the pure getData() but ensures that the data are ordered (oldest at the 
+         * beginning, newest at the end of the vector
+         * This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+         * :return new linearized vector
+        */
+        std::vector<T> getLinearizedData() const;
+
+        bool operator==(const SlidingWindow& r2) const;
+        bool operator!=(const SlidingWindow& r2) const;
+        T& operator[](UInt index);
+        const T& operator[](UInt index) const;
+
+
+      private:
+        std::vector<T> buffer_;
+        UInt idxNext_;
+        size_t size_;
+}; 
+}} //end ns
+
+/// IMPLEMENTATION
+#include <cmath>
+
+using nupic::util::SlidingWindow;
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(nupic::UInt maxCapacity) :
+  maxCapacity(maxCapacity) {
+  NTA_CHECK(maxCapacity > 0);
+  buffer_.reserve(maxCapacity);
+  buffer_.resize(maxCapacity);
+  idxNext_ = 0;
+  size_ = 0;
+}
+
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(nupic::UInt maxCapacity, std::vector<T> initialData)  :
+  SlidingWindow(maxCapacity) {
+  auto sz = std::min(initialData.size(), (size_t)maxCapacity);
+  buffer_.insert(std::begin(buffer_), std::end(initialData) - sz, std::end(initialData));
+  buffer_.resize(sz);
+  idxNext_ = sz % maxCapacity;
+  size_ = sz;
+}
+
+
+template<class T>
+size_t SlidingWindow<T>::size() const {
+  NTA_ASSERT(buffer_.size() <= maxCapacity);
+  return std::min(size_, (size_t)maxCapacity);
+}
+
+
+template<class T>
+void SlidingWindow<T>::append(T newValue) {
+    buffer_[idxNext_] = newValue;
+    idxNext_ = ++idxNext_ %maxCapacity;//the assignment must be out of the [] above -not [idxNext_++%maxCap]- because we want to store the value %maxCap, not only ++
+  ++size_;
+}
+
+
+template<class T>
+T SlidingWindow<T>::append(T newValue, bool& isValid) {
+  T old = buffer_.empty() ? newValue : buffer_[idxNext_]; //handle case of empty buffer (access buff[0]), otherwise return oldest elem
+  isValid = (buffer_.size()==maxCapacity); //only in this case we drop oldest; this happens always after first maxCap steps ; must be checked before append()
+  append(newValue);
+  return old; 
+}
+
+
+template<class T>
+const std::vector<T>& SlidingWindow<T>::getData() const {
+  return buffer_; //may contain trailing "zeros"
+}
+
+
+template<class T>
+std::vector<T> SlidingWindow<T>::getLinearizedData() const {
+  std::vector<T> lin;
+  lin.reserve(buffer_.size());
+
+  lin.insert(std::begin(lin), std::begin(buffer_) + idxNext_, std::end(buffer_)); //insert the "older" part at the beginning
```

Comment:
It's a nitpick, but I really think the current is better. With your approach, if you exchange the 2 `insert()` lines, the code won't work, so the current is more self-descriptive: We really want to rewrite `lin` from the beginning, even if it weren't empty.